atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java

```diff
@@ -194,39 +199,76 @@ void markRequestConnectionException() {
+    private CassandraClientPool(CassandraKeyValueServiceConfig config,
+            StartupChecks startupChecks,
+            ScheduledExecutorService refreshDaemon,
+            ExecutorService initExecutor) {
         this.config = config;
-        config.servers().forEach(this::addPool);
-        refreshDaemon = Tracers.wrap(PTExecutors.newScheduledThreadPool(1, new ThreadFactoryBuilder()
-                .setDaemon(true)
-                .setNameFormat("CassandraClientPoolRefresh-%d")
-                .build()));
-        refreshDaemon.scheduleWithFixedDelay(() -> {
-            try {
-                refreshPool();
-            } catch (Throwable t) {
-                log.error("Failed to refresh Cassandra KVS pool."
-                        + " Extended periods of being unable to refresh will cause perf degradation.", t);
+        this.refreshDaemon = refreshDaemon;
+        this.startupChecks = startupChecks;
+        this.initExecutor = initExecutor;
+        isInitialised.compareAndSet(false, false);
+        initialize();
+    }
+
+    @Override
+    public void initialize() {
+        initExecutor.execute(() -> {
+            while (!isInitialised.get()) {
+                initWithoutThrowing();
             }
-        }, config.poolRefreshIntervalSeconds(), config.poolRefreshIntervalSeconds(), TimeUnit.SECONDS);
+        });
+    }
+
+    private void initWithoutThrowing() {
+        try {
+            config.servers().forEach(this::addPool);
+            refreshDaemon.scheduleWithFixedDelay(() -> {
+                try {
+                    refreshPool();
+                } catch (Throwable t) {
+                    log.error("Failed to refresh Cassandra KVS pool."
+                                    + " Extended periods of being unable to refresh will cause perf degradation.",
+                            t);
+                }
+            }, config.poolRefreshIntervalSeconds(), config.poolRefreshIntervalSeconds(), TimeUnit.SECONDS);
 
-        // for testability, mock/spy are bad at mockability of things called in constructors
-        if (startupChecks == StartupChecks.RUN) {
-            runOneTimeStartupChecks();
+            // for testability, mock/spy are bad at mockability of things called in constructors
+            if (startupChecks == StartupChecks.RUN) {
+                runOneTimeStartupChecks();
+            }
+            refreshPool(); // ensure we've initialized before returning
+            registerAggregateMetrics();
+            isInitialised.compareAndSet(false, true);
+            log.info("Initialization of the ClientPool was successful.");
+        } catch (Exception ex) {
+            cleanResources();
+            log.error(
+                    "Failed to initialize the Cassandra clientPool most likely because Cassandra is not healthy, "
+                            + "retrying after 5 seconds.");
+            Uninterruptibles.sleepUninterruptibly(5, TimeUnit.SECONDS);
```

Comment:
this retry logic could probably be moved up to the caller, to keep this method focused on initialization